plugins/ssh/ssh.plugin.zsh

```diff
@@ -0,0 +1,39 @@
+############################################################
+# Take all host sections in .ssh/config and offer them for
+# completion as hosts (e.g. for ssh, rsync, scp and the like)
+# Filter out wildcard host sections.
+local hosts
+if [[ -f $HOME/.ssh/config ]]; then
+  hosts=($(egrep '^Host.*' $HOME/.ssh/config | awk '{print $2}' | grep -v '^*' | sed -e 's/\.*\*$//'))
+  zstyle ':completion:*:hosts' hosts $hosts
+fi
+
+############################################################
+# Remove host key from known hosts based on a host section
+# name from .ssh/config
+function ssh_rmhkey {
+  if [[ "x$1" == "x" ]]; then return; fi
+  ssh-keygen -R $(grep -A10 $1 ~/.ssh/config | grep -i HostName | head -n 1 | awk '{print $2}')
+}
+compctl -k hosts ssh_rmhkey
+
+############################################################
+# Load SSH key into agent
+function ssh_load_key() {
+  local key=$1
+  if [[ "$key" == "" ]]; then return; fi
+  if ( ! ssh-add -l | grep -q $key ); then
+    ssh-add ~/.ssh/$key;
```

Comment:
and quote the whole `"~/.ssh/$key"`

Reply:
This won't work, as quoting `~` makes it not expand to `$HOME`.

Reply:
Wow, that's true. I'd expect to expand in double quotes, and do not in single quotes, which is the rationale of not expanding in any quotes?

Reply:
Tilde expansion is a separate kind that works different than parameter expansion, the rule about double and single quotes only applies to dollar ($) strings. You can look up `tilde expansion` on the zsh docs.

Reply:
The bigger problem here is, this won't _work_. If `$key` is a private key filename under `$HOME/.ssh/`, its path won't be visible in the output of `ssh-add -l` if that fingerprint has already been unlocked. The output consists of lines that look like this:

```text
2048 SHA256:fingerprintfingerprint comment (RSA)
```

(Where "comment" is frequently "user@remotehost".)

So, this is going to be doing a lot of unnecessary scanning of `ssh-add -l` output, looking for filenames that aren't displayed.

I'd say it would be better to just remove the conditional, save one invocation each of `ssh-add` and `grep`, and just always call `ssh-add` unconditionally, except that `ssh-add /path/to/private_key` for a private key file that requires a passphrase will **always** prompt for that passphrase, even if the key is already added and unlocked.

But the right way to get the fingerprint for a private key is to run `ssh-keygen -l -f /path/to/private_key`. The output of **that** command is the same fingerprint line that will be listed in `ssh-add -l` if the key is added and unlocked.